### ISSUE TITLE
using pydantic causes stdlib dataclass frozen+slots behavior to change (read-only attribute error).

### ISSUE THREAD

**Issue by RossBencina** [pydantic/pydantic#6840]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

I am observing the following unexpected behavior...

The following code works fine, it does not use pydantic:

```
from dataclasses import dataclass

@dataclass(slots=True, frozen=True)
class MyStdLibDataclass:
    x : int = 0

d1 = MyStdLibDataclass(x=100)
assert d1.x == 100
```
However, if I append the following code to the same file:

```
from pydantic import BaseModel

class MyModel(BaseModel):
    y : MyStdLibDataclass|None = None

d2 = MyStdLibDataclass(x=101) # error here
assert d2.x == 101
```

I get the error:
```
Traceback (most recent call last):
  File "C:\Users\Ross\Desktop\dataclass_bug.py", line 15, in <module>
    d2 = MyStdLibDataclass(x=101)
  File "<string>", line 3, in __init__
AttributeError: 'MyStdLibDataclass' object attribute 'x' is read-only
```

### Example Code

```Python
from dataclasses import dataclass
from pydantic import BaseModel

@dataclass(slots=True, frozen=True)
class MyStdLibDataclass:
    x : int = 0

d1 = MyStdLibDataclass(x=100) # ok
assert d1.x == 100

class MyModel(BaseModel):
    y : MyStdLibDataclass|None = None

d2 = MyStdLibDataclass(x=101) # error here
assert d2.x == 101
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.0.2
        pydantic-core version: 2.1.2 release build profile
                 install path: C:\Python310\Lib\site-packages\pydantic
               python version: 3.10.7 (tags/v3.10.7:6cc6b13, Sep  5 2022, 14:08:36) [MSC v.1933 64 bit (AMD64)]
                     platform: Windows-10-10.0.19045-SP0
     optional deps. installed: ['typing-extensions']
```

Selected Assignee: @davidhewitt

**Comment by hramezani** [comment]
I think it's a duplicate of https://github.com/pydantic/pydantic/issues/6783

It got fixed in https://github.com/pydantic/pydantic/commit/9fb218be17e2aa4372fb0c815d5f6be5fc3046a9 and will be released in next release(soon)

**Comment by RossBencina** [comment]
@hramezani: I just installed your main branch from git and it does indeed appear to resolve the error message.

What does "next release(soon)" mean? Next few days? weeks? I was hoping to merge Pydantic integration in my project by the end of the week but this is obviously a showstopper.

Also, is there some documentation about why pydantic is messing with stdlib stuff? I had assumed that if I use stdlib dataclasses I would be 100% isolated from pydantic. I know that my example embeds a stdlib dataclass in a BaseModel subclass, but I also observed the issue with stdlib dataclasses that are not so embedded (these cases too are fixed by your main branch).
